src/backend/distributed/transaction/transaction_management.c

```diff
@@ -0,0 +1,142 @@
+/*-------------------------------------------------------------------------
+ *
+ * transaction_management.c
+ *
+ *   Transaction management for Citus.  Most of the work is delegated to other
+ *   subsystems, this files, and especially CoordinatedTransactionCallback,
+ *   coordinates the work between them.
+ *
+ * Copyright (c) 2016, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "libpq-fe.h"
+
+#include "miscadmin.h"
+
+#include "access/xact.h"
+#include "distributed/connection_management.h"
+#include "distributed/hash_helpers.h"
+#include "distributed/transaction_management.h"
+#include "utils/hsearch.h"
+
+
+CoordinatedTransactionState CurrentCoordinatedTransactionState = COORD_TRANS_NONE;
+
+/* GUC, the commit protocol to use for commands affecting more than one connection */
+int MultiShardCommitProtocol = COMMIT_PROTOCOL_1PC;
+
+/* state needed to keep track of operations used during a transaction */
+XactModificationType XactModificationLevel = XACT_MODIFICATION_NONE;
+
+
+static bool subXactAbortAttempted = false;
+
+
+/* transaction management functions */
+static void CoordinatedTransactionCallback(XactEvent event, void *arg);
+static void CoordinatedSubTransactionCallback(SubXactEvent event, SubTransactionId subId,
+											  SubTransactionId parentSubid, void *arg);
+
+
+void
+InitializeTransactionManagement(void)
+{
+	/* hook into transaction machinery */
+	RegisterXactCallback(CoordinatedTransactionCallback, NULL);
+	RegisterSubXactCallback(CoordinatedSubTransactionCallback, NULL);
+}
+
+
+/*
+ * Transaction management callback, handling coordinated transaction, and
+ * transaction independent connection management.
+ *
+ * NB: There should only ever be a single transaction callback in citus, the
+ * ordering between the callbacks and thee actions within those callbacks
+ * otherwise becomes too undeterministic / hard to reason about.
+ */
+static void
+CoordinatedTransactionCallback(XactEvent event, void *arg)
+{
+	switch (event)
+	{
+		case XACT_EVENT_COMMIT:
+		{
+			/* close connections etc. */
+			if (CurrentCoordinatedTransactionState != COORD_TRANS_NONE)
+			{
+				AtEOXact_Connections(true);
+			}
+
+			Assert(!subXactAbortAttempted);
+			CurrentCoordinatedTransactionState = COORD_TRANS_NONE;
+			XactModificationLevel = XACT_MODIFICATION_NONE;
+		}
+		break;
+
+		case XACT_EVENT_ABORT:
+		{
+			/* close connections etc. */
+			if (CurrentCoordinatedTransactionState != COORD_TRANS_NONE)
+			{
+				AtEOXact_Connections(false);
+			}
+
+			CurrentCoordinatedTransactionState = COORD_TRANS_NONE;
+			XactModificationLevel = XACT_MODIFICATION_NONE;
+			subXactAbortAttempted = false;
+		}
+		break;
+
+		case XACT_EVENT_PARALLEL_COMMIT:
+		case XACT_EVENT_PARALLEL_ABORT:
+		case XACT_EVENT_PREPARE:
+		{ }
```

Comment:
Should we error out on prepare if there is a coordinated transaction until we actually implement it? (if possible)